### ISSUE TITLE
Multi-Release-Epic: Sharding

### ISSUE THREAD

**Issue by sttts** [kcp-dev/kcp#1225]
# Demo Objective
- [ ] kcp admin should be able to add shards for added capacity, each with its own etcd server.
- [ ] shards should keep working (to be defined what this means exactly) when other shards die or are unavailable.
- [x] https://github.com/kcp-dev/kcp/pull/1203 the front-proxy knows how to forward workspace requests to shards
- add (hard-coded) replication of certain object between shards:
  - [ ] defining ClusterWorkspace `X` in the parent workspace `parent(X)` on the shard the workspace `X` lives on.
  - [ ] APIExport that are bound by at least one non-local binding

# Action Items

- [x] https://github.com/kcp-dev/kcp/pull/1203 forward requests to the right shard from the front-proxy
- [x] https://github.com/kcp-dev/kcp/pull/1203 sharded-test-server binary
- [x] https://github.com/kcp-dev/kcp/pull/1203 setup single-shard + front-proxy CI job
- [x] @p0lyn0mial verify where we are with watch cache and performant label selector

========= WIP, but won't merge by v0.6 =========

- [x] @p0lyn0mial hardcode second informer with non-local data.
- [x] @p0lyn0mial prototype and experiment ~with informer merging, and~ how to enable existing controller to work with second informer
- [x] get 2-shard e2e-sharded into place in CI, with limited test coverage
- get ClusterWorkspace, APIExports, APIBindings, auth to work
  - [x] @p0lyn0mial APIBindings on shards can bind APIExport on other shard
  - [x] @p0lyn0mial ClusterWorkspaces can be scheduled to second shard

========= up to here for v0.7 ==========

#### add identities configmap in system:shard @p0lyn0mial 

- [x] https://github.com/kcp-dev/kcp/pull/1725: implement a controller that maintains a local cm with export identities
- [x] https://github.com/kcp-dev/kcp/pull/1727 update the identity provider to read data from a local config map
- [x] https://github.com/kcp-dev/kcp/pull/1781 wire the APIIdentityCacheController to the kcp server

#### implement replication controller / cache server
   - [x] @p0lyn0mial  create initial `cmd/cache-server` and `pkg/cache/server/{config.go,server.go}` and `pkg/cache/server/options` plumbing (https://github.com/kcp-dev/kcp/pull/1790)
     - follow options/config/server pattern tightly
     - options should be compatible and easily composable with those in pkg/server/options 
   - [x] @p0lyn0mial  instantiate `apiextensions-apiserver` and wire CRD lister similar to today's `pkg/server/apiextensions.go` with at least APIExports and APIResourceSchemas
     - use empty schemas, as dumb as possible: APIExport/APIResourceSchema CRDs can be static as in https://github.com/kcp-dev/kcp/blob/8d2ed6a6ba467ddb4dcba0a7e1675cf8382223c3/pkg/virtual/apiexport/schemas/apis.go#L36
     - no admission, no validation, no conversion, no pruning, no SSA (if possible), for now no authorization
     - PRs: https://github.com/kcp-dev/kcp/pull/1811, https://github.com/kcp-dev/kcp/pull/1815, https://github.com/kcp-dev/kubernetes/pull/93

========= up to here for v0.8 ==========

#### implement replication controller / cache server (continued)

   - [x] @p0lyn0mial wire custom etcd prefix for "cached CRDs" returned by the CRD lister: `/registry/group/cache:resource:identity:shard/…` (https://github.com/kcp-dev/kcp/pull/1841)
   - [x] @p0lyn0mial assign shard names to objects (https://github.com/kcp-dev/kubernetes/pull/100, https://github.com/kcp-dev/kcp/pull/1926, https://github.com/kcp-dev/kcp/pull/1928)
   - [x] @p0lyn0mial serve `/services/cache/shards/*/clusters/*` (https://github.com/kcp-dev/kcp/pull/1961)
   - [x] @p0lyn0mial serve `/services/cache/shards/{shard-name}/clusters/{cluster-name}` (https://github.com/kcp-dev/kcp/pull/1961)
   - [ ] @p0lyn0mial instantiate cache-server inside of `kcp` binary. (https://github.com/kcp-dev/kcp/pull/1949, https://github.com/kcp-dev/kcp/pull/1947, https://github.com/kcp-dev/kcp/pull/1954)
   - [x] @p0lyn0mial wire `--cache-url` into `kcp`, defaulting to localhost. (https://github.com/kcp-dev/kcp/pull/1970)
   - [ ] @p0lyn0mial create shard controller pushing (labeled to be published) apiexports, apiresourceschemas, ClusterWorkspaces to cache
   - [ ] create cache controller creating "cached CRDs" or create "cached CRD" on the fly in the CRD lister for all cached APIExports
   - [ ] serve internal APIs through schema-less CRDs, wired statically
   - [ ] wire secondary informers to cache instead of the other shard
   - [ ] enable the watch-cache

#### authorization
   - [ ] shards should only be able to write their shard partition of the cache

#### subscription model
   - [ ] implement subscription model @sttts 
  
  ```golang
  type CachingInformer interface {
      Informer

      KeepWatching(clusterName logicalcluster.Name, labelSelector, ttl time.Time)
  }
  ```
#### data removal
unused resources could be automatically removed

#### shard management, a few loose ideas: 
what will decide when a new shard needs to be created ?
which cluster it gets deployed to ?
making other shards aware of the new shard ?

#### (**in parallel, best effort**) add some other controller(s), e.g. ClusterWorkspaces

[ ] wire informer both ways @p0lyn0mial 
   - this includes a controller to cope with not-yet-synced informers
   
 [ ] start the second informer (aka. `TemporaryRootShardKcpSharedInformerFactory`) in a dedicated post-start-hook @p0lyn0mial 

#### Testing
- [ ] get all e2e green with two shards

## Backlog:
- ThingReplicationClaim
- prototype CRDB service
- field selectors
- authz of replication
- controller programming model, related to #1339
- [ ] https://github.com/kcp-dev/kcp/issues/1339
- [ ] https://github.com/kcp-dev/kcp/issues/342
- [x] https://github.com/kcp-dev/kcp/issues/354
- [ ] https://github.com/kcp-dev/kcp/issues/348

# Stories
- CRDB implements ThingReplicationClaim
- etcd implements ThingReplicationClaim

# Context
- ThingReplicationClaim https://docs.google.com/document/d/10-HCQPwLSXfs3-6I-_dKNXdenTeJ3rrx8ZTbX00X-Nc/edit#

**Comment by stevekuznetsov** [comment]
Superseded by #1999
